lib/dns/resource.js

```diff
@@ -83,369 +78,76 @@ const DUMMY = Buffer.alloc(0);
 class Resource extends Struct {
   constructor() {
     super();
-    this.version = 0;
-    this.ttl = 0;
-    this.compat = false;
-    this.hosts = [];
-    this.canonical = null;
-    this.delegate = null;
-    this.ns = [];
-    this.service = [];
-    this.uri = [];
-    this.email = [];
-    this.text = [];
-    this.location = [];
-    this.magnet = [];
-    this.ds = [];
-    this.tls = [];
-    this.smime = [];
-    this.ssh = [];
-    this.pgp = [];
-    this.addr = [];
-    this.extra = [];
+    this.ttl = DEFAULT_TTL;
+    this.records = [];
   }
 
-  compress() {
-    const c = new Compressor();
-
-    for (const host of this.hosts)
-      host.compress(c);
-
-    if (this.canonical)
-      this.canonical.compress(c);
-
-    if (this.delegate)
-      this.delegate.compress(c);
-
-    for (const ns of this.ns)
-      ns.compress(c);
-
-    for (const srv of this.service)
-      srv.compress(c);
-
-    for (const uri of this.uri)
-      c.add(uri);
-
-    for (const email of this.email)
-      c.add(email);
-
-    for (const text of this.text)
-      c.add(text);
-
-    for (const loc of this.location)
-      loc.compress(c);
-
-    for (const urn of this.magnet)
-      urn.compress(c);
-
-    for (const ds of this.ds)
-      ds.compress(c);
+  hasType(type) {
+    assert((type & 0xff) === type);
 
-    for (const tls of this.tls)
-      tls.compress(c);
-
-    for (const smime of this.smime)
-      smime.compress(c);
+    for (const record of this.records) {
+      if (record.type === type)
+        return true;
+    }
 
-    for (const ssh of this.ssh)
-      ssh.compress(c);
+    return false;
+  }
 
-    for (const pgp of this.pgp)
-      pgp.compress(c);
+  hasNS() {
+    for (const {type} of this.records) {
+      if (type < hsTypes.NS || type > hsTypes.SYNTH6)
+        continue;
 
-    for (const addr of this.addr)
-      addr.compress(c);
+      return true;
+    }
 
-    for (const extra of this.extra)
-      extra.compress(c);
+    return false;
+  }
 
-    return c;
+  hasDS() {
+    return this.hasType(hsTypes.DS);
   }
 
   encode() {
```

Comment:
We could use `MAX_RESOURCE_SIZE` here to replace the magic number `512`.
https://github.com/handshake-org/hsd/blob/af86d4899791601d10212115c80a2f2136fdb004/lib/covenants/rules.js#L107

Not part of this PR, but it could also be used here:
https://github.com/handshake-org/hsd/blob/8252c97e890148c56903c3178c9d3869318a285e/lib/covenants/namestate.js#L838